pkg/lint/lint.go

```diff
@@ -0,0 +1,180 @@
+				errors = append(
+					errors,
+					fmt.Sprintf(
+						"path %q in %q is a duplicate of the same path in config.yaml",
+						path, overshadowingPatch.filepath,
+					),
+				)
+			}
+		}
+	}
+	return errors
+}
+
+func GlobalOvershadowedValues(d *Discovery) (errors []string) {
+	if len(d.Installations) == 0 {
+		return // avoid division by 0
+	}
+	for path, valuePath := range d.Config.paths {
+		if float64(len(valuePath.OvershadowedBy)/len(d.Installations)) >= overshadowErrorThreshold {
+			errors = append(
+				errors,
+				fmt.Sprintf(
+					"path %q in config.yaml is overshadowed by %d/%d patches; consider removing it from config.yaml",
+					path, len(valuePath.OvershadowedBy), len(d.Installations),
+				),
+			)
+		}
+	}
+	return errors
+}
+
+func PatchUnusedValues(d *Discovery) (errors []string) {
+	for _, configPatch := range d.ConfigPatches {
+		if len(d.AppsPerInstallation[configPatch.installation]) == 0 {
+			continue // avoid division by 0
+		}
+		for path, valuePath := range configPatch.paths {
+			if len(valuePath.UsedBy) == 0 {
+				errors = append(
+					errors,
+					fmt.Sprintf(
+						"path %q in %q is *unused*; consider removing it",
+						path, configPatch.filepath,
+					),
+				)
+			} else if float64(len(valuePath.UsedBy)/len(d.AppsPerInstallation[configPatch.installation])) <= patchUsedByErrorThreshold {
+				errors = append(
+					errors,
+					fmt.Sprintf(
+						"path %q in %q is used by %d/%d apps; consider moving it to app templates",
+						path, configPatch.filepath, len(valuePath.UsedBy), len(d.AppsPerInstallation[configPatch.installation]),
+					),
+				)
+			}
+		}
+	}
+	return errors
+}
+
+func GlobalConfigUnusedValues(d *Discovery) (errors []string) {
+	if len(d.Installations) == 0 || len(d.Apps) == 0 {
+		return // what's the point, nothing is defined
+	}
+	for path, valuePath := range d.Config.paths {
+		if len(valuePath.UsedBy) == 0 {
+			errors = append(
+				errors,
+				fmt.Sprintf(
+					"path %q in %q is *unused*; consider removing it",
+					path, d.Config.filepath,
+				),
+			)
+		} else if len(valuePath.UsedBy) == 1 {
+			errors = append(
+				errors,
+				fmt.Sprintf(
+					"path %q in %q is used by just one app in %q; consider moving it",
+					path, d.Config.filepath, valuePath.UsedBy[0].filepath,
+				),
+			)
+		}
+	}
+	return errors
+}
+
+// TODO: linter func might need to return internal errors instead of panicking!
+func UnusedPatchableAppValues(d *Discovery) (errors []string) {
+	for _, template := range d.Templates {
+		for path, value := range template.values {
+			if !value.MayBeMissing {
+				continue
+			}
+
+			used := false
+			for _, templatePatch := range d.TemplatePatches {
+				// render template with zero data
+				t := templatePatch.CopyTemplate().Option("missingkey=zero")
+				output := bytes.NewBuffer([]byte{})
+				var data interface{}
+				err := t.Execute(output, data)
+				if err != nil {
+					panic(err)
```

Comment:
TODO: I'm not handling this error properly, because I want to throw this chunk of code in `discovery.go`.